### PR TITLE
Dirty hierarchy

### PR DESCRIPTION
**Pull request by Gabo-7**
# [CPF-3111](https://trippinc.atlassian.net/browse/CPF-3111)
- [x] Updated Version
- [x] Updated Change Log
# Description
Added testing for dirty transforms and hierarchies.
A major refactor by @mbracks was also done.
# Testing
Please create a checklist of what to test for.
- [ ] Test that dirty transforms and dirty hierarchies are being flagged correctly.
- [x] Test the UV infractions still work correctly

[CPF-3111]: https://trippinc.atlassian.net/browse/CPF-3111?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mbracks** [comment]
Lex Imperialis finds Flipped UVs and Overlapping UVs on clean Pheonix
- :white_check_mark: UV infraction
image

**Comment by mbracks** [comment]
- :white_check_mark: Dirty Hierarchy check finds empty game objects
- :x: Dirty transforms doesn't find dirty transforms on objects that were exported as a group

File: Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs [hunk 1]
```diff
@@ -8,144 +10,195 @@ public class ModelJudicator : ImporterJudicator
...
+                else
+                    return result;
+            }
+            else
+            {
+                if (hasSecondaryUVSet)
+                {
+                    bool hasFlippedUVs2 = HasFlippedUVs(mesh.uv2, mesh.triangles, mesh);
+                    bool hasOverlappingUVs2 = HasOverlappingTriangles(mesh, mesh.uv2);
+                    result.Add(new Infraction
```
Comment (mbracks) [inline Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs:105]: You need an if statement to check the results of the bools above before adding a new infraction.

File: Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs [hunk 2]
```diff
@@ -8,144 +10,195 @@ public class ModelJudicator : ImporterJudicator
...
+                if(rootObject.transform.localPosition != Vector3.zero)
+                result = new Infraction
                 {
                     isFixable = false,
-                    message = $"{objectName} is missing a secondary UV set for lightmaps."
-                });
+                    message = $"{rootObject.name}: The pivot is not set at the origin (0,0,0)"
+                };
+            }
+            else if(rootObject.GetComponents<Component>().Count() == 1)
```
Comment (mbracks) [inline Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs:186]: In this scenario, the root object isn't checked for dirty transforms only it's children.

File: Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs [hunk 3]
```diff
@@ -8,144 +10,195 @@ public class ModelJudicator : ImporterJudicator
...
+            {
+                message += " [ Overlapping UVs ]";
             }
+            return message;
+        }
 
-            if (hasOverlappingUVs1)
+        private List<Infraction> CheckForEmptyNodeInfactions(ModelImporter modelImporter)
+        {
+            List<Infraction> result = new List<Infraction>();
```
Comment (mbracks) [inline Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs:140]: Initializing the list early.

File: Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs [hunk 4]
```diff
@@ -8,144 +10,195 @@ public class ModelJudicator : ImporterJudicator
...
-            return judgment; // Return the populated judgment
+            return result;
         }
 
-        private Mesh GetMesh(GameObject gameObject)
+        private List<Mesh> GetSubMeshes(ModelImporter modelImporter)
         {
-            // Attempt to get the MeshFilter component from the provided GameObject
-            if (gameObject.TryGetComponent(out MeshFilter meshFilter))
+            List<Mesh> subMeshes = new List<Mesh>();
```
Comment (mbracks) [inline Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs:64]: Early initialization of a list.

**Review by mbracks — CHANGES_REQUESTED** [review]

File: Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs (hunk 3 above)
Comment (Gabo-7) [reply Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs:140]: The initialization is necessary. Not initializing, or using null, makes it break; compile error and runtime error respectively.

File: Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs (hunk 1 above)
Comment (Gabo-7) [reply Packages/com.tripp.leximperialis/Editor/ModelJudicator.cs:105]: Added a check, but some things are still unclear

**Review by mbracks — CHANGES_REQUESTED** [review]

**Review by mbracks — APPROVED** [review]
Great work

**Pull request merged**